nutils/evaluable.py

```diff
@@ -5243,6 +5321,556 @@ def eval_sparse(funcs: AsEvaluableArray, **arguments: typing.Mapping[str, numpy.
+    #     └ e1: LoopSum          'j'->(0,0)   (0,1)
+    #       └ e2: Add                         (0,0,0)
+    #         ├ e3: LoopIndex    'j'->(0,0)   (0,0,0)
+    #         │ └ e4: Constant                (0,)
+    #         └ e5: LoopIndex    'i'->(0,)    (0,0)
+    #           └ e6: Constant                (0,)
+    #
+    # Note that `e4` and `e6` are the lengths of the loops. In the second
+    # example loop `e9` does not depend on loop index `i`, but is a dependency
+    # of loop `e7`. Therefor, `e9` gets loop id `(0,)` and `e7` id `(1,)`
+    # (adjacent to `e9`):
+    #
+    #                            loop_id      block_id
+    #     e7: LoopSum            'i'->(1,)    (2,)
+    #     └ e8: Add                           (1,0)
+    #       ├ e9: LoopSum        'j'->(0,)    (1,)
+    #       │ └ e10: LoopIndex   'j'->(0,)    (0,0)
+    #       │   └ e4: Constant                (0,)
+    #       └ e11: LoopIndex     'i'->(1,)    (1,0)
+    #         └ e6: Constant                  (0,)
+    #
+    # Having defined the loop and block ids, the loops (without content) are
+    # serialized. For the first example this gives
+    #
+    #     def compiled(**a):
+    #         # contents of blocks[0,]
+    #         for v5 in range(c6):
+    #             # contents of blocks[0,0]
+    #             for v3 in range(c4):
+    #                 # contents of blocks[0,0,0]
+    #             # contents of blocks[0,1]
+    #         # contents of blocks[1,]
+    #         return v0
+    #
+    # The final step is the serialization of the `Evaluable`s. Each `Evaluable`
+    # compiles first its dependencies, then itself, placing the statements in
+    # the block with the corresponding block id. The serialization of the first
+    # example (without inplace additions):
+    #
+    #     def compiled(**a):
+    #         v0 = numpy.zeros((), int) # e0 init
+    #         for v5 in range(c6):
+    #             v1 = numpy.zeros((), int) # e1 init
+    #             for v3 in range(c4):
+    #                 v2 = v3 + v5
+    #                 v1 += v2 # e1 update
+    #             # e1 exit
+    #             v0 += v1 # e0 update
+    #         # e0 exit
+    #         return v0
+
+    if stats is None:
+        stats = 'log' if graphviz else False
+    elif stats not in ('log', False):
+        raise ValueError(f'`stats` must be `None`, `False` or `"log"` but got {stats!r}')
+
+    # Build return value format string `ret` with the same structure as `func`
+    # and convert `func` to a flat list.
+    stack = [func]
+    ret_fmt = []
+    funcs = []
+    MakeTuple = collections.namedtuple('MakeTuple', ('n'))
+    while stack:
+        obj = stack.pop(0)
+        if isinstance(obj, MakeTuple):
+            m = len(ret_fmt) - obj.n
+            ret_fmt[m:] = ['(' + ', '.join(ret_fmt[m:]) + (',)' if obj.n == 1 else ')')]
+        elif isinstance(obj, (tuple, list)):
+            stack[:0] = [*obj, MakeTuple(len(obj))]
```

Comment:
Wouldn't it be more natural to use a right sided stack, so that you can use `append` and `extend`?

Reply:
Fixed.